### ISSUE TITLE
bug: MPI_Ibarrier failure on OSX with gcc-6.2 from brew

### ISSUE THREAD

**Issue by balay** [pmodels/mpich#2437]
We are seeing a failure with a PETSc test code on OSX with gcc-6.2 from brew. This code is valgrind clean on linux [with mpich].

Hoping you might be able to reproduce this with MPI test suite

```
* thread #1: tid = 0x1f5a146, 0x00007fffb9022506 libdyld.dylib`stack_not_16_byte_aligned_error, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=EXC_I386_GPFLT)
  * frame #0: 0x00007fffb9022506 libdyld.dylib`stack_not_16_byte_aligned_error
    frame #1: 0x00007fff56311eb0
    frame #2: 0x000000010bddf849 libpmpi.12.dylib`MPID_Sched_start(sp=0x00007fff56311f20, comm=0x000000010bea88c0, tag=31, req=0x00007fff56311f30) + 965 at mpid_sched.c:424
    frame #3: 0x000000010bbe1c33 libpmpi.12.dylib`MPIR_Ibarrier_impl(comm_ptr=0x000000010bea88c0, request=0x00007fff56312054) + 651 at ibarrier.c:203
    frame #4: 0x000000010b992a9b libmpi.12.dylib`MPI_Ibarrier(comm=-2080374780, request=0x00007fff56312054) + 1343 at ibarrier.c:288
    frame #5: 0x00000001099bd9a1 libpetsc.3.07.dylib`PetscCommBuildTwoSided_Ibarrier(comm=-2080374780, count=1, dtype=1275069445, nto=2, toranks=0x00007fb036018060, todata=0x00007fb036016860,
nfrom=0x00007fb03601a864, fromranks=0x00007fb03601a868, fromdata=0x00007fff56312158) + 2279 at mpits.c:139
    frame #6: 0x00000001099bfd22 libpetsc.3.07.dylib`PetscCommBuildTwoSided(comm=-2080374780, count=1, dtype=1275069445, nto=2, toranks=0x00007fb036018060, todata=0x00007fb036016860,
nfrom=0x00007fb03601a864, fromranks=0x00007fb03601a868, fromdata=0x00007fff56312158) + 1016 at mpits.c:298
    frame #7: 0x0000000109b902f7 libpetsc.3.07.dylib`PetscSFSetUp_Basic(sf=0x00007fb036015e60) + 996 at sfbasic.c:552
    frame #8: 0x0000000109b9d352 libpetsc.3.07.dylib`PetscSFSetUp(sf=0x00007fb036015e60) + 916 at sf.c:192
    frame #9: 0x0000000109ba7d50 libpetsc.3.07.dylib`PetscSFReduceBegin(sf=0x00007fb036015e60, unit=1275069445, leafdata=0x00007fb03600f660, rootdata=0x00007fb03600b660, op=1476395015) + 1177 at sf.c:1034
    frame #10: 0x000000010a32c3b6 libpetsc.3.07.dylib`MatZeroRowsMapLocal_Private(A=0x00007fb035044c60, N=20, rows=0x00007fb03600f660, nr=0x00007fff563122fc, olrows=0x00007fff56312300) + 1833 at
zerorows.c:42
    frame #11: 0x000000010a253a0e libpetsc.3.07.dylib`MatZeroRows_MPIAIJ(A=0x00007fb035044c60, N=20, rows=0x00007fb03600f660, diag=0, x=0x0000000000000000, b=0x0000000000000000) + 514 at mpiaij.c:792
    frame #12: 0x0000000109d544cf libpetsc.3.07.dylib`MatZeroRows(mat=0x00007fb035044c60, numRows=20, rows=0x00007fb03600f660, diag=0, x=0x0000000000000000, b=0x0000000000000000) + 1468 at matrix.c:5824
    frame #13: 0x00000001098ef84e ex3`main(argc=3, args=0x00007fff563125f8) + 4415 at ex3.c:132
    frame #14: 0x00007fffb9026255 libdyld.dylib`start + 1
    frame #15: 0x00007fffb9026255 libdyld.dylib`start + 1

```
@BarrySmith @jedbrown

**Comment by BarrySmith** [comment]
The code seems to run fine with the Xcode clang compilers.

**Comment by knepley** [comment]
Does anyone even read these? Should we just stop using MPICH? This is 9 months on. Kids have been born who were not even conceived when we reported this bug.

**Comment by rtmills** [comment]
I have run into the same issues with GCC 7.1.0 (installed via HomeBrew on my Mac running Mac OS 10.12.5). Using Apple's provided clang compilers, things work fine. Would be great if this worked with the GCC suite, however.

**Comment by raffenet** [comment]
@balay is this still reproducible with the latest master branch? This commit (https://github.com/pmodels/mpich/commit/c52614d40f9a141a77bddd9b32a6f501854d6d6f) fixed a bunch of issues on OSX having to do with alignment.

**Comment by balay** [comment]
I see this issue with latest mpich master a5abbc071
```
balay@ipro^~/petsc/arch-gcc(master) $ cd externalpackages/mpich-git/
balay@ipro^~/petsc/arch-gcc/externalpackages/mpich-git(master) $ git show -q --oneline
a5abbc071 (HEAD -> master, origin/master, origin/HEAD) maint: adds bootstrap.sh script that downloads/compiles/installs dependencies.
balay@ipro^~/petsc/arch-gcc/externalpackages/mpich-git(master) $ ../../bin/mpichversion 
MPICH Version:    	3.3a2
MPICH Release date:	unreleased development copy
MPICH Device:    	ch3:sock
MPICH configure: 	--prefix=/Users/balay/petsc/arch-gcc MAKE=/usr/bin/make --libdir=/Users/balay/petsc/arch-gcc/lib CC=gcc-7 CFLAGS=-fstack-protector -g3 AR=/usr/bin/ar ARFLAGS=cr CXX=g++-7 CXXFLAGS=-fstack-protector -g F77=gfortran FFLAGS=-ffree-line-length-0 -g FC=gfortran FCFLAGS=-ffree-line-length-0 -g --enable-shared --with-device=ch3:sock --with-pm=hydra --enable-fast=no --enable-g=meminit
MPICH CC: 	gcc-7 -fstack-protector -g3   -O0
MPICH CXX: 	g++-7 -fstack-protector -g 
MPICH F77: 	gfortran -ffree-line-length-0 -g 
MPICH FC: 	gfortran -ffree-line-length-0 -g 
MPICH Custom Information: 	
```

**Comment by raffenet** [comment]
OK, thanks for confirming. We will take a look.

**Comment by raffenet** [comment]
@balay can you try #2718 and see if this issue goes away? I'm not sure if I'm treating the symptom or the disease, but it fixed what I was able to reproduce in my test environment.

**Comment by balay** [comment]
Tried  raffenet:sched-dump -  the examples don't crash anymore

**Comment by balay** [comment]
@raffenet I don't see this fix in newly released mpich-3.2.1. Will there be  mpich-3.3a3 or similar release with the fix?

**Comment by raffenet** [comment]
Hmm, you are right. We must have scanned past it because it didn't have "fix" in the commit message. We have discussed doing another bug-fix release in the 3.2 series soon because of a Slurm integration issue in Hydra. We could include this, as well.

Otherwise, yes, we will have a 3.3 series release (probably beta) out sometime before end of the year.

**Comment by raffenet** [comment]
@balay actually, does this issue occur with the 3.2.x series? The sched code has been shifted around significantly in master, making the patches in #2718 hard to apply.

**Comment by balay** [comment]
Yes the problem exists in 3.2.x.
I had tried 3.2.1 last week - and could see this issue.

**Comment by balay** [comment]
Sure I can skip 3.2 and use the  3.3 beta tarball.

Alternative is to use a 3.3 snapshot tarball - which I'm trying to avoid..

**Comment by raffenet** [comment]
Fixed in 3.3. Closing.
